protocols/autonat/src/behaviour.rs

```diff
@@ -0,0 +1,638 @@
+                                    peer_id: peer,
+                                    handler: self.inner.new_handler(),
+                                    condition: DialPeerCondition::Always,
+                                });
+                            }
+                            None => {
+                                let response = DialResponse::Err(ResponseError::DialRefused);
+                                let _ = self.inner.send_response(channel, response);
+                            }
+                        },
+                        RequestResponseMessage::Response {
+                            request_id: _,
+                            response,
+                        } => {
+                            if let Some(status) = self.handle_response(peer, Some(response)) {
+                                self.reachability = Reachability::from(&status);
+                                self.last_probe = Instant::now();
+                                return Poll::Ready(NetworkBehaviourAction::GenerateEvent(status));
+                            }
+                        }
+                    },
+                    RequestResponseEvent::ResponseSent { .. } => {}
+                    RequestResponseEvent::OutboundFailure { peer, .. } => {
+                        if let Some(status) = self.handle_response(peer, None) {
+                            self.reachability = Reachability::from(&status);
+                            self.last_probe = Instant::now();
+                            return Poll::Ready(NetworkBehaviourAction::GenerateEvent(status));
+                        }
+                    }
+                    RequestResponseEvent::InboundFailure { peer, .. } => {
+                        self.ongoing_inbound.remove(&peer);
+                    }
+                },
+                Poll::Ready(action) => return Poll::Ready(action.map_out(|_| unreachable!())),
+                Poll::Pending => return Poll::Pending,
+            }
+            if let Some(probe) = self.pending_probe.take() {
+                let records: Vec<_> = params.external_addresses().collect();
+                self.do_probe(probe, records)
+            } else if let Some(auto_retry) = self.auto_retry.as_ref() {
+                if Instant::now().duration_since(self.last_probe) > auto_retry.interval {
+                    let probe = self.new_probe(
+                        auto_retry.required_success,
+                        Vec::new(),
+                        true,
+                        auto_retry.extend_with_connected,
+                        auto_retry.max_peers,
+                        auto_retry.min_valid,
+                    );
+                    let records: Vec<_> = params.external_addresses().collect();
+                    self.do_probe(probe, records);
+                }
+            }
+        }
+    }
+}
+
+// Filter demanded dial addresses for validity, to prevent abuse.
+fn filter_valid_addrs(
```

Comment:
:rocket: 

Could you add a unit test to ensure all IP addresses in `demanded` are replaced with the IP address in `observed_remote_at`?

Reply:
Added in ebd9d2a4d8b6f606a8ee5f3192a03e153743b9d0.